File: web/src/api/health.ts
import { request } from './client';

/**
 * Mirrors `ReadinessResponse` in the API's `schemas/health.py`.
 *
 * Hand-written for now, and that is a known seam: two declarations of one contract drift. When the
 * API surface grows past a couple of endpoints, generate these from the OpenAPI schema FastAPI
 * already publishes at `/openapi.json` rather than maintaining them by hand.
 */
export interface DependencyStatus {
  readonly healthy: boolean;
  readonly detail: string | null;
}

/**
 * The guest purge, as `/health/ready` reports it (ADR-0019, slice 1.6).
 *
 * Seven fields: a flag, an instant, two counts, a word, a staleness verdict and a detail. **Nothing
 * here identifies anybody** — no session id, no filename, no storage key, no path (AC-36) — and
 * nothing here can change `ready` or the status code (AC-32); `jobs` is read, `dependencies`
 * decides.
 *
 * `detail` carries an *exception type name* when a field could not be computed, and this client
 * deliberately never renders it — see `features/health/purgeView.ts`, which does not carry it onto
 * the view at all.
 */
export interface GuestPurgeStatus {
  /** Whether a schedule fires this job at all (`GUEST_PURGE_ENABLED`). */
  readonly scheduled: boolean;
  /** RFC 3339, UTC, whole-second. `null` when it has never run, or the heartbeat was flushed. */
  readonly last_run: string | null;
  /** `null` exactly when `last_run` is. */
  readonly last_run_age_seconds: number | null;
  readonly last_outcome: 'ok' | 'failed' | null;
  /** Expired sessions still waiting. **`null` when the count could not be taken** (R-29). */
  readonly overdue: number | null;
  /** `scheduled && (last_run is null || age > 3h)`. `scheduled: false` implies `false` (AC-33). */
  readonly stale: boolean;
  /** An exception *type*, only when a field above could not be computed. Never rendered. */
  readonly detail: string | null;
}

export interface Readiness {
  readonly ready: boolean;
  readonly dependencies: Record<string, DependencyStatus>;
  /**
   * The scheduled jobs this deployment reports.
   *
   * **Both `jobs` and `guest_purge` are optional here and *required* in the API's Pydantic model,
   * and that asymmetry is deliberate rather than an oversight** (R-32). It reads as an
   * inconsistency, so: this API always sends the member, but during a deploy a browser can be
   * served a bundle *newer* than the API process answering it. The two declarations of this one
   * contract are versioned independently — one is in an image, one is on a CDN — so the client
   * types what it might actually receive, not what the current server promises. An older API with
   * no `jobs` member must render the *empty* state ("not reported by this API"), which is a
   * sentence an operator can act on, rather than throwing on a missing property.
   *
   * The server, having no such uncertainty, never omits it. Widening the *server* type to match
   * would be the wrong fix: it would make a field that is always present look optional to every
   * construction site, which is how a fact nobody computed gets published as a confident default.
   */
  readonly jobs?: {
    readonly guest_purge?: GuestPurgeStatus;
  };
}

/**
 * Fetch the readiness report.
 *
 * 503 is accepted rather than thrown: an unready API answers 503 *with* the report naming which
 * dependency is down, and that report is precisely what the caller asked for.
 */
export function fetchReadiness(signal?: AbortSignal): Promise<Readiness> {
  return request<Readiness>('/health/ready', {
    acceptStatuses: [503],
    ...(signal ? { signal } : {}),
  });
}

File: api/src/tailorcraft/infrastructure/api/schemas/health.py
"""Response schemas for the health endpoints."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class DependencyStatus(BaseModel):
    """One probed dependency."""

    healthy: bool
    detail: str | None = None


class GuestPurgeStatus(BaseModel):
    """The guest purge, reported as a fact (ADR-0019, AC-31).

    **The key set is the contract and a test pins it.** Seven fields, in the order the ADR writes
    them; this model is the wire shape of `health.probes.JobStatus` and nothing more.

    **Nothing in here can change the status code or `ready`** — the safety lives in the *type* one
    layer down (`JobStatus` has no `healthy` field, so `all(r.healthy …)` cannot reach it), not in a
    promise made here. This model is why the field names are stable; that model is why they are
    harmless.

    **No field has a default.** Every one is supplied explicitly at the single construction site, on
    the house rule `PurgeReport` states: a default on a reported fact is how a second construction
    site publishes a confident `false` for something it never computed. `stale` is the sharp one —
    it is the field an operator acts on.

    **`jobs` is required in this response and optional in the TypeScript client**, and the asymmetry
    is deliberate (R-32): this API always sends it, and during a deploy a browser can be served a
    bundle newer than the API it talks to. The client renders "not reported"; the server never
    omits.

    Nothing here identifies anybody: a flag, an instant, a count, a word, and an exception's type
    (Constitution §8, AC-16, AC-36).
    """

    scheduled: bool
    last_run: str | None
    last_run_age_seconds: int | None
    last_outcome: Literal["ok", "failed"] | None
    overdue: int | None
    stale: bool
    detail: str | None


class JobsStatus(BaseModel):
    """The scheduled jobs this deployment reports.

    A named model rather than `dict[str, …]` (which is what `dependencies` is, one field up) because
    the members do not share a shape: a dependency is always `{healthy, detail}`, while a job
    publishes whatever facts *that* job has. ADR-0019 decision 6 says the two existing beat sweeps
    land here when they grow a heartbeat, and each will bring its own fields — a dict would force
    them into one type or into `Any`.

    **A new member here is never a new way to fail readiness** (ADR-0019 decision 6). That rule is
    worth more than this class: it is the reason the next person adding a job does not have to
    re-litigate the 503 question.
    """

    guest_purge: GuestPurgeStatus


class ReadinessResponse(BaseModel):
    """The readiness report.

    Deliberately names every dependency it checked, rather than collapsing to a single boolean. A
    bare `{"status": "ok"}` tells an operator at 2 a.m. nothing they can act on, and — worse — looks
    identical whether it probed three dependencies or none.

    Since slice 1.6 it carries two *different kinds* of thing, and the split is the point:
    `dependencies` decides `ready` and the status code; `jobs` decides nothing and is read
    (ADR-0019). **A green `/health/ready` no longer means "everything is fine"** — it means this
    process can serve a request. Reading `jobs.guest_purge` is a separate act.
    """

    ready: bool
    dependencies: dict[str, DependencyStatus] = Field(default_factory=dict)
    jobs: JobsStatus


class LivenessResponse(BaseModel):
    """The liveness report: this process is running. It claims nothing else."""

    alive: bool = True
